Fix doctest code-blocks in documentation

vchuravy opened the pull request:
Due to backtrace and printing changes most of the doctests are currently failing. 

Remaining issues are:
### Changed syntax errors

```
File "manual/variables.rst", line 131, in default
Failed example:
    try = "No"
Expected:
    ERROR: syntax: unexpected "="
Got:
    ERROR: syntax: unhandled expr (error #<julia: "unexpected "="">)
     in eval(::Module, ::Any) at ./boot.jl:237
```
### Method overwritten

```
File "manual/stacktraces.rst", line 173, in default
Failed example:
    @noinline parent() = child()
Expected:
    parent (generic function with 1 method)
Got:
    WARNING: Method definition parent() in module Main at none:1 overwritten at none:1.
    parent (generic function with 1 method)
```
### Changed Unit-range printing

```
File "manual/arrays.rst", line 312, in default
Failed example:
    searchsorted(a, 3)
Expected:
    3:2
Got:
    0-element UnitRange{Int64}
```
### Inf in Tuple printing

```
Failed example:
    (typemin(Float16),typemax(Float16))
Expected:
    (-Inf16,Inf16)
Got:
    (-Inf,Inf)
```
## Todo
- [x] Decide if unitrange is a regression (if not reformulate example)
- [x] Are syntax error a regression? (They seem to be less readable than before.)
- [x] Enable doctests on travis
- [x] Fix doctests in stdlib
- [x] Fix doctests in devdocs
- [x] What to do with method overwritten errors (They are clutter and don't add to the examples, on the other hand renaming the methods is excessive)

vchuravy committed:
fix most doctests in the manual that changed due to backtraces.

yuyichao commented:
Ref https://github.com/JuliaLang/julia/issues/13966

yuyichao commented:
> What to do with method overwritten errors

I think we can just rename the method....

vchuravy committed:
fix stdlib doctests

vchuravy committed:
fix doctests in devdocs

vchuravy commented:
@nalimilan, was this intended to be a doctest?

```
File "manual/documentation.rst", line 42, in default
Failed example:
    bar([1, 2], [1, 2])
Expected:
    1
    ```
    """
    function bar(x, y) ...
Got:
    ERROR: UndefVarError: bar not defined
     in eval(::Module, ::Any) at ./boot.jl:237
```

Otherwise we are down to 40 failures all in the style described above. And where I am not quite certain how to proceed.

vchuravy committed:
resolve Warning mismatch by renaming functions

vchuravy commented:
I renamed the problematic functions, but I am open to better suggestions. 

```
Doctest summary
===============
 1056 tests
    7 failures in tests
    0 failures in setup code
    0 failures in cleanup code
```

vchuravy commented:
I think we can merge this and I will open issues about the remaining failing tests.

nalimilan commented:
That `bar` code block wasn't supposed to be a doctest. I think I kept how things were before without thinking too much about it.

vchuravy committed:
convert superflous doctest to code-block

vchuravy commented:
Thanks fixed.

vchuravy committed:
Fix example about typemax and typemin

Due to the changes in 4706184a tuples are now printed compact. Thus to
demonstrate the point in this example we have to execute the examples
independently.

vchuravy committed:
fixup linenumbers

vchuravy commented:
All done. All tests are now passing

```
Doctest summary
===============
 1058 tests
    0 failures in tests
    0 failures in setup code
    0 failures in cleanup code
build succeeded.
```

tkelman commented on doc/devdocs/types.rst:
```diff
@@ -194,7 +194,7 @@ One can construct :obj:`TypeVar`\s manually:
 
    julia> TypeVar(:V, Signed, Real, false)
-   Signed<:V<:Real
+   V<:Real
```
what happened here?

yuyichao replied:
https://github.com/JuliaLang/julia/commit/fff2b9aa811549186ec39e71cc4f6331203e2119 should the `in_env` be `false` if the `:tvar_env` context is not found?

tkelman replied:
who should we be asking that question?

yuyichao replied:
https://github.com/JuliaLang/julia/pull/16221

tkelman commented on doc/manual/integers-and-floating-point-numbers.rst:
```diff
@@ -412,14 +412,23 @@ types:
 .. doctest::
 
-    julia> (typemin(Float16),typemax(Float16))
-    (-Inf16,Inf16)
```
was the tuple not displaying correctly?

yuyichao replied:
I think somehow the function we use to show a tuple changed.

```jl
julia> show(Inf16)
Inf16
julia> print(Inf16)
Inf
```

IIRC this also affects the precision of floating point numbers in tuple. (there should be at least one other broken doctest due to this)

vchuravy replied:
The call sequence that happens is:
```
writemime(STDOUT, MIME"text/plain"(), (Inf16, Inf16)))

writemime(io::IO, ::MIME"text/plain", x) =  showcompact(io, x)
showcompact(io, x) = show(IOContext(STDOUT, :limit_output => true), (Inf16, Inf16))
show(io::IO, t::Tuple) = show_delim_array(io, t, '(', ',', ')', true)
```
`show_delim_array` finally calls `show(io, x)`, but `io = IOContext(STDOUT, :limit_output => true)`
and that calls `showcompact(io, x)`, witch calls `Base.showcompact(io::IO, x::Float16) = _show(io, x, PRECISION, 5, false)`, where the `false` signifies that the number should be printed non-typed.

Now the question is if it is the right behaviour for `showcompact` to drop the type information.

tkelman commented on doc/manual/types.rst:
```diff
@@ -432,7 +434,7 @@ You can also change the values as one would expect:
 
     julia> foo.qux = 2
-    2.0
+    2
```
what caused this? it's a Float64 field

yuyichao replied:
This one makes sense, since the value of an assignment is always the original value

vchuravy replied:
In Julia v0.4
```
julia> a = foo.qux = 2
2.0
julia> typeof(a)
Float64
```
In v0.5
```
julia> a = foo.qux = 2
2

julia> typeof(a)
Int64
```

yuyichao replied:
I believe this is a bug on 0.4 https://github.com/JuliaLang/julia/issues/14691 . It's a breaking change so I'm not sure if it should be backported.

yuyichao commented on doc/stdlib/collections.rst:
```diff
@@ -1331,17 +1318,11 @@ inserted and priorities accessed or changed using indexing notation.
-      "c" => 15
-      "b" => 5
-      "a" => 0
+    Base.Collections.PriorityQueue{Any,Any,Base.Order.ForwardOrdering}("c"=>15,"b"=>5,"a"=>0)
```
I believe this is https://github.com/JuliaLang/julia/issues/14684

tkelman commented:
Bump - this was great work, but goes stale quickly. Can you rebase and see where we stand?

yuyichao committed:
start->first

vchuravy commented:
I won't have time for a rebase till next week, but I will take a look after that.

omus commented:
@vchuravy could you use assistance with this PR? I was hoping to add a new doctest.

vchuravy commented:
@omus I just rebased, but there is an problem with Array printing that causes doctest to fail. I don't have time to look into it right now so feel free to give it a go.

tkelman commented on base/docs/helpdb/Base.jl:
```diff
@@ -5643,7 +5643,8 @@ julia> deleteat!([6, 5, 4, 3, 2, 1], 1:2:5)
 ERROR: ArgumentError: indices must be unique and sorted
- in deleteat! at array.jl:543
+ in deleteat!(::Array{Int64,1}, ::Tuple{Int64,Int64}) at ./array.jl:565
+ in eval(::Module, ::Any) at ./boot.jl:237
```
I feel like the doctest system should be pruning this line

omus commented:
I'm currently working on fixing doctests. Expect a PR before the end of JuliaCon

tkelman commented:
I'm going to close this in favor of #17106 then, thanks to both @vchuravy and @omus
